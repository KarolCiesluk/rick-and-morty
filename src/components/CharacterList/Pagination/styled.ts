import styled, { css } from "styled-components";

import { ReactComponent as Arrow } from "./arrow.svg";

export const StyledPagination = styled.nav`
  text-align: center;
  margin-top: 40px;
`;

export const LeftArrow = styled(Arrow)`
  transform: rotate(180deg);
`;

export const PaginationButton = styled.button`
  background: none;
  border: none;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  color: #9B7954;

  ${({ disabled }) => disabled && css`
    color: gray;
    cursor: not-allowed;
  `}
`;

export const PageNumber = styled.span`
  margin: 0 6px;
`;