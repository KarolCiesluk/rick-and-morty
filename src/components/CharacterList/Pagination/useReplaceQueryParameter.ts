import { useSearchParams } from "react-router-dom";

interface KeyValuePair {
  key: string;
  value: string | undefined;
}

export const useReplaceQueryParameter = () => {
  const [searchParams, setSearchParams] = useSearchParams();

  return ({ key, value }: KeyValuePair) => {
    if (!value) {
      searchParams.delete(key);
    } else {
      searchParams.set(key, value);
    }

    setSearchParams(searchParams);
  };
};