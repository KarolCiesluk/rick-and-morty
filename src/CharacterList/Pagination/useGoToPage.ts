import { useReplaceQueryParameter } from "./useReplaceQueryParameter";

export const useGoToPage = () => {
  const replaceQueryParameter = useReplaceQueryParameter();

  return (pageNumber: number) => replaceQueryParameter({
    key: "page",
    value: pageNumber === 1 ? undefined : pageNumber.toString(),
  });
};