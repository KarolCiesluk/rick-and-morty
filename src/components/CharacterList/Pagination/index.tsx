import { useGoToPage } from "./useGoToPage";
import { ReactComponent as Arrow } from "./arrow.svg";
import {
  LeftArrow,
  PageNumber,
  PaginationButton,
  StyledPagination
} from "./styled";

interface PaginationProps {
  lastPage: number;
  currentPage: number;
}

const Pagination = ({ lastPage, currentPage }: PaginationProps) => {
  const goToPage = useGoToPage();

  const goToFirstPage = () => {
    goToPage(1);
  };

  const goToPreviousPage = () => {
    goToPage(currentPage - 1);
  };

  const goToNextPage = () => {
    goToPage(currentPage + 1);
  };

  const goToLastPage = () => {
    goToPage(lastPage);
  };

  return (
    <StyledPagination>

      <PaginationButton
        disabled={currentPage === 1}
        onClick={goToFirstPage}
      >
        <LeftArrow /><LeftArrow />
      </PaginationButton>

      <PaginationButton
        disabled={currentPage === 1}
        onClick={goToPreviousPage}
      >
        <LeftArrow />
      </PaginationButton>

      <PageNumber>{currentPage} of {lastPage}</PageNumber>

      <PaginationButton
        disabled={currentPage === lastPage}
        onClick={goToNextPage}
      >
        <Arrow />
      </PaginationButton>

      <PaginationButton
        disabled={currentPage === lastPage}
        onClick={goToLastPage}
      >
        <Arrow /><Arrow />
      </PaginationButton>

    </StyledPagination>
  );
};

export default Pagination;