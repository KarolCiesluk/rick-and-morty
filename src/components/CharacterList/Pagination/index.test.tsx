import { mount } from "enzyme";
import { HashRouter } from "react-router-dom";
import renderer from "react-test-renderer";

import Pagination from ".";

describe("Pagination testing", () => {
  it("renders ok", () => {
    const tree = renderer
      .create(
        <HashRouter>
          <Pagination lastPage={826} currentPage={1} />
        </HashRouter>
      ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("should render '2 of 826' in span html tag", () => {
    const wrapper = mount(
      <HashRouter>
        <Pagination lastPage={826} currentPage={2} />
      </HashRouter>
    );

    expect(wrapper.find("span").text())
      .toBe("2 of 826");
  });

  it("should render 4 buttons", () => {
    const wrapper = mount(
      <HashRouter>
        <Pagination lastPage={826} currentPage={2} />
      </HashRouter>
    );

    expect(wrapper.find("button").length).toBe(4);
  });

  it("should button 'back to first page' and button 'go to previous page' be disabled when current page equals to 1", () => {
    const wrapper = mount(
      <HashRouter>
        <Pagination lastPage={826} currentPage={1} />
      </HashRouter>
    );

    const firstPageButton = wrapper.find("button").get(0);
    expect(firstPageButton.props.disabled).toBe(true);

    const previousPageButton = wrapper.find("button").get(1);
    expect(previousPageButton.props.disabled).toBe(true);
  });

  it("should button 'go to last page' and button 'go to next page' be enabled when current page equals to 1", () => {
    const wrapper = mount(
      <HashRouter>
        <Pagination lastPage={826} currentPage={1} />
      </HashRouter>
    );

    const nextPageButton = wrapper.find("button").get(2);
    expect(nextPageButton.props.disabled).toBe(false);

    const lastPageButton = wrapper.find("button").get(3);
    expect(lastPageButton.props.disabled).toBe(false);
  });

  it("should button 'back to first page' and button 'go to previous page' be enabled when current page equals to 826", () => {
    const wrapper = mount(
      <HashRouter>
        <Pagination lastPage={826} currentPage={826} />
      </HashRouter>
    );

    const firstPageButton = wrapper.find("button").get(0);
    expect(firstPageButton.props.disabled).toBe(false);

    const previousPageButton = wrapper.find("button").get(1);
    expect(previousPageButton.props.disabled).toBe(false);
  });

  it("should button 'go to last page' and button 'go to next page' be disabled when current page equals to 826", () => {
    const wrapper = mount(
      <HashRouter>
        <Pagination lastPage={826} currentPage={826} />
      </HashRouter>
    );

    const nextPageButton = wrapper.find("button").get(2);
    expect(nextPageButton.props.disabled).toBe(true);

    const lastPageButton = wrapper.find("button").get(3);
    expect(lastPageButton.props.disabled).toBe(true);
  });

  it("should all buttons be enabled when current page equals to 5", () => {
    const wrapper = mount(
      <HashRouter>
        <Pagination lastPage={826} currentPage={5} />
      </HashRouter>
    );

    const firstPageButton = wrapper.find("button").get(0);
    expect(firstPageButton.props.disabled).toBe(false);

    const previousPageButton = wrapper.find("button").get(1);
    expect(previousPageButton.props.disabled).toBe(false);

    const nextPageButton = wrapper.find("button").get(2);
    expect(nextPageButton.props.disabled).toBe(false);

    const lastPageButton = wrapper.find("button").get(3);
    expect(lastPageButton.props.disabled).toBe(false);
  });
});