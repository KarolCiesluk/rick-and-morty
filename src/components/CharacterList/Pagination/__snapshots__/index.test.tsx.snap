// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination testing renders ok 1`] = `
<nav
  className="sc-bdvvtL zVTrh"
>
  <button
    className="sc-dkPtRN cFPYvG"
    disabled={true}
    onClick={[Function]}
  >
    <svg
      className="sc-gsDKAQ huTkWC"
    >
      arrow.svg
    </svg>
    <svg
      className="sc-gsDKAQ huTkWC"
    >
      arrow.svg
    </svg>
  </button>
  <button
    className="sc-dkPtRN cFPYvG"
    disabled={true}
    onClick={[Function]}
  >
    <svg
      className="sc-gsDKAQ huTkWC"
    >
      arrow.svg
    </svg>
  </button>
  <span
    className="sc-hKwDye NlkbS"
  >
    1
     of 
    826
  </span>
  <button
    className="sc-dkPtRN jYQpi"
    disabled={false}
    onClick={[Function]}
  >
    <svg>
      arrow.svg
    </svg>
  </button>
  <button
    className="sc-dkPtRN jYQpi"
    disabled={false}
    onClick={[Function]}
  >
    <svg>
      arrow.svg
    </svg>
    <svg>
      arrow.svg
    </svg>
  </button>
</nav>
`;

exports[`Pagination tests renders ok 1`] = `
<nav
  className="sc-bdvvtL zVTrh"
>
  <button
    className="sc-dkPtRN cFPYvG"
    disabled={true}
    onClick={[Function]}
  >
    <svg
      className="sc-gsDKAQ huTkWC"
    >
      arrow.svg
    </svg>
    <svg
      className="sc-gsDKAQ huTkWC"
    >
      arrow.svg
    </svg>
  </button>
  <button
    className="sc-dkPtRN cFPYvG"
    disabled={true}
    onClick={[Function]}
  >
    <svg
      className="sc-gsDKAQ huTkWC"
    >
      arrow.svg
    </svg>
  </button>
  <span
    className="sc-hKwDye NlkbS"
  >
    1
     of 
    826
  </span>
  <button
    className="sc-dkPtRN jYQpi"
    disabled={false}
    onClick={[Function]}
  >
    <svg>
      arrow.svg
    </svg>
  </button>
  <button
    className="sc-dkPtRN jYQpi"
    disabled={false}
    onClick={[Function]}
  >
    <svg>
      arrow.svg
    </svg>
    <svg>
      arrow.svg
    </svg>
  </button>
</nav>
`;
